### PR TITLE
{Role} Fix: `az role assignment list-changelogs` fails with KeyError

### PR DESCRIPTION
**Pull request by jiasli**
## Symptom

`az role assignment list-changelogs` fails with `KeyError`:

```
> az role assignment list-changelogs --start-time 2020-07-21T12:29:00Z --end-time 2020-07-21T12:30:00Z
The command failed with an unexpected error. Here is the traceback:

'/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/aro-nw3napmz/providers/Microsoft.Authorization/roleDefinitions/b24988ac-6180-42a0-ab88-20f7382dd24c'
Traceback (most recent call last):
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-_kvuguic\knack\knack\cli.py", line 215, in invoke
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure\cli\core\commands\__init__.py", line 654, in execute
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure\cli\core\commands\__init__.py", line 718, in _run_jobs_serially
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure\cli\core\commands\__init__.py", line 711, in _run_job
  File "C:\Users\VSSADM~1\AppData\Local\Temp\pip-install-_kvuguic\six\six.py", line 703, in reraise
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure\cli\core\commands\__init__.py", line 688, in _run_job
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure\cli\core\commands\__init__.py", line 325, in __call__
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure\cli\core\__init__.py", line 545, in default_command_handler
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure\cli\command_modules\role\custom.py", line 334, in list_role_assignment_change_logs
KeyError: '/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/aro-nw3napmz/providers/Microsoft.Authorization/roleDefinitions/b24988ac-6180-42a0-ab88-20f7382dd24c'
```

## Cause

This is because when the role assignment is created, it uses a **child** role definition 

    /subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/aro-nw3napmz/providers/Microsoft.Authorization/roleDefinitions/b24988ac-6180-42a0-ab88-20f7382dd24c
  
which inherits from the **actual** role definition

    /subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590                            /providers/Microsoft.Authorization/roleDefinitions/b24988ac-6180-42a0-ab88-20f7382dd24c

This can be verified by calling [Role Definitions - Get](https://docs.microsoft.com/en-us/rest/api/authorization/roledefinitions/get):

```
> az rest -u "https://management.azure.com/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/aro-nw3napmz/providers/Microsoft.Authorization/roleDefinitions/b24988ac-6180-42a0-ab88-20f7382dd24c?api-version=2015-07-01"
{
  "id": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/providers/Microsoft.Authorization/roleDefinitions/b24988ac-6180-42a0-ab88-20f7382dd24c",
  "name": "b24988ac-6180-42a0-ab88-20f7382dd24c",
  "properties": {
    ...
    "roleName": "Contributor",
    "type": "BuiltInRole",
    "updatedBy": null,
    "updatedOn": "2019-02-05T21:24:38.4580610Z"
  },
  "type": "Microsoft.Authorization/roleDefinitions"
}
```

but the result from [Role Definitions - List](https://docs.microsoft.com/en-us/rest/api/authorization/roledefinitions/list) only has the **actual** role definition, not the **child** one, so the key lookup fails to retrieve the **child** one:

https://github.com/Azure/azure-cli/blob/4c28536a6db8c8e43b7182d0aaf202ac2b4ab777/src/azure-cli/azure/cli/command_modules/role/custom.py#L334

This PR changes the lookup logic to **only use the resource name (last GUID part) as role definition key**, so that the role definition can be resolved and shown correctly:

```
> az role assignment list-changelogs --start-time 2020-07-21T12:29:00Z --end-time 2020-07-21T12:30:00Z
[
  {
    "action": "Granted",
    "caller": "3e675290-31e1-49f6-a3cd-03df3d7d6e2e",
    "principalId": "a62417b1-489d-4e8c-b7b3-c5e32ca6774a",
    "principalName": "https://az.aro.azure.com/09a2399a-59c4-4953-b08b-e9d052e49f93",
    "roleDefinitionId": "b24988ac-6180-42a0-ab88-20f7382dd24c",
    "roleName": "Contributor",
    "scope": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourcegroups/aro-nw3napmz",
    "scopeName": "aro-nw3napmz",
    "scopeType": "Resource group",
    "timestamp": "2020-07-21T12:29:56.561242+00:00"
  },
  ...
]
```

## Command Usage

Luckily, according to client telemetry, in the last 7 days, there are only 29 invocations of this command from 8 subscriptions. 

## Remarks

This issue is discovered when fixing the live tests for `role` command module (#14317).

### CODE REVIEW

**Comment by yonzhan** [comment]
RBAC

**Comment by jiasli** [comment]
Merged into https://github.com/Azure/azure-cli/pull/14317.

File: src/azure-cli/azure/cli/command_modules/role/custom.py [hunk 1]
```diff
@@ -280,8 +280,12 @@ def list_role_assignment_change_logs(cmd, start_time=None, end_time=None):  # py
     result = []
     worker = MultiAPIAdaptor(cmd.cli_ctx)
     start_events, end_events, offline_events, client = _get_assignment_events(cmd.cli_ctx, start_time, end_time)
-    role_defs = {d.id: [worker.get_role_property(d, 'role_name'),
-                        d.id.split('/')[-1]] for d in list_role_definitions(cmd)}
+
+    # Use the resource `name` of roleDefinitions as keys, instead of `id`, because `id` can be inherited.
+    #   name: b24988ac-6180-42a0-ab88-20f7382dd24c
+    #   id: /subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/providers/Microsoft.Authorization/roleDefinitions/b24988ac-6180-42a0-ab88-20f7382dd24c  # pylint: disable=line-too-long
+    role_defs = {d.name: [worker.get_role_property(d, 'role_name'), d.id.split('/')[-1]]
```
Comment (arrownj) [inline src/azure-cli/azure/cli/command_modules/role/custom.py]: Are the names returned by `list_role_definitions` always different ?

Comment (jiasli) [reply src/azure-cli/azure/cli/command_modules/role/custom.py]: Yes. It is a UUID and always unique.

**Review by arrownj — APPROVED** [review]
👍

File: src/azure-cli/azure/cli/command_modules/role/custom.py (hunk 1 above)
Comment (jiasli) [reply src/azure-cli/azure/cli/command_modules/role/custom.py]: As tested, creating a `RoleDefinition` on sub-resource will always fallback to the subscription:

```
> az rest -m PUT -u https://management.azure.com/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1/providers/Microsoft.Authorization/roleDefinitions/05675f70-84a5-44b5-94bf-2a6354b8077d?api-version=2018-01-01-preview -b "@role-definition-rest.json"
{
  "id": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/providers/Microsoft.Authorization/roleDefinitions/05675f70-84a5-44b5-94bf-2a6354b8077d",
  "name": "05675f70-84a5-44b5-94bf-2a6354b8077d",
...
```

<details><summary> role-definition-rest.json </summary>

```
{
    "properties": {
        "roleName": "Contoso On-call 4",
        "type": "CustomRole",
        "description": "Perform VM actions and read storage and network information.",
        "assignableScopes": [
            "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1"
        ],
        "permissions": [
            {
                "actions": [
                    "Microsoft.Compute/*/read",
                    "Microsoft.Compute/virtualMachines/start/action",
                    "Microsoft.Compute/virtualMachines/restart/action",
                    "Microsoft.Network/*/read",
                    "Microsoft.Storage/*/read",
                    "Microsoft.Authorization/*/read",
                    "Microsoft.Resources/subscriptions/resourceGroups/read",
                    "Microsoft.Resources/subscriptions/resourceGroups/resources/read",
                    "Microsoft.Insights/alertRules/*",
                    "Microsoft.Support/*"
                ],
                "notActions": [],
                "dataActions": [
                    "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/*"
                ],
                "notDataActions": [
                    "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write"
                ]
            }
        ]
    },
    "type": "Microsoft.Authorization/roleDefinitions"
}
```

</details>

Notice that `id` doesn't have `/resourceGroups/rg1`.

File: src/azure-cli/azure/cli/command_modules/role/custom.py [hunk 2]
```diff
@@ -331,8 +334,12 @@ def list_role_assignment_change_logs(cmd, start_time=None, end_time=None):  # py
                         else:
                             entry['scopeType'] = 'Resource'
 
-                    entry['roleDefinitionId'] = role_defs[payload['roleDefinitionId']][1]
-                    entry['roleName'] = role_defs[payload['roleDefinitionId']][0]
+                    # Look up the resource `name`, like b24988ac-6180-42a0-ab88-20f7382dd24c
+                    role_resource_name = payload['roleDefinitionId'].split('/')[-1]
+                    entry['roleDefinitionId'] = role_resource_name
```
Comment (jsntcy) [inline src/azure-cli/azure/cli/command_modules/role/custom.py:339]: Who will use entry['roleDefinitionId'] later? Now it is like `b24988ac-6180-42a0-ab88-20f7382dd24c`, is it expected?

Comment (jiasli) [reply src/azure-cli/azure/cli/command_modules/role/custom.py:339]: This key has the same value as the original behavior:

```
d.id.split('/')[-1]
```

I didn't change the value.

**Review by jsntcy — APPROVED** [review]

**Pull request merged**